src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp

```diff
@@ -0,0 +1,142 @@
+#include <Storages/MergeTree/ParallelReplicasReadingCoordinator.h>
+
+#include <algorithm>
+#include <vector>
+#include <compare>
+#include <numeric>
+#include <unordered_map>
+#include <map>
+#include <iostream>
+#include <set>
+#include <cassert>
+
+
+#include <base/logger_useful.h>
+#include <base/types.h>
+#include <base/scope_guard.h>
+#include <Common/Stopwatch.h>
+#include "IO/WriteBufferFromString.h"
+#include <Storages/MergeTree/MarkRange.h>
+#include <Storages/MergeTree/IntersectionsIndexes.h>
+
+namespace DB
+{
+
+class ParallelReplicasReadingCoordinator::Impl
+{
+public:
+
+    using PartitionReadRequestPtr = std::unique_ptr<PartitionReadRequest>;
+
+    using PartToMarkRanges = std::map<String, HalfIntervals>;
+
+    struct PartitionReading
+    {
+        PartSegments part_ranges;
+        PartToMarkRanges mark_ranges_in_part;
+    };
+
+    using PartitionToBlockRanges = std::map<String, PartitionReading>;
+    PartitionToBlockRanges partitions;
+
+    std::mutex mutex;
+
+    PartitionReadResponse handleRequest(PartitionReadRequest request);
+};
+
+
+PartitionReadResponse ParallelReplicasReadingCoordinator::Impl::handleRequest(PartitionReadRequest request)
+{
+    AtomicStopwatch watch;
+    std::lock_guard lock(mutex);
+
+    auto partition_it = partitions.find(request.partition_id);
+
+    SCOPE_EXIT({
+        LOG_TRACE(&Poco::Logger::get("ParallelReplicasReadingCoordinator"), "Time for handling request: {}ns", watch.elapsed());
+    });
+
+    /// We are the first who wants to process parts in partition
+    if (partition_it == partitions.end())
+    {
+        auto part_and_projection = request.part_name + "#" + request.projection_name;
```

Comment:
Why don't we store part name and projection separately?

Reply:
I want to underline their strict connectivity and also this concatenation simplifies the code a little bit (There are no additional checks that we want to read from the same projection). Maybe we can move them to std::pair or somewhat similar.